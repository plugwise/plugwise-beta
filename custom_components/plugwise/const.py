"""Constants for Plugwise beta component."""
from datetime import timedelta
import logging
from typing import Final

import voluptuous as vol
from homeassistant.const import Platform
from homeassistant.helpers import config_validation as cv

DOMAIN: Final = "plugwise"

LOGGER = logging.getLogger(__package__)

API: Final = "api"
ATTR_ENABLED_DEFAULT: Final = "enabled_default"
COORDINATOR: Final = "coordinator"
CONF_COOLING_ON: Final = "cooling_on"
CONF_HOMEKIT_EMULATION: Final = "homekit_emulation"  # pw-beta
CONF_REFRESH_INTERVAL: Final = "refresh_interval"  # pw-beta
CONF_MANUAL_PATH: Final = "Enter Manually"
GATEWAY: Final = "gateway"
ID: Final = "id"
PW_LOCATION: Final = "location"
PW_TYPE: Final = "plugwise_type"
SMILE: Final = "smile"
STICK: Final = "stick"
STRETCH: Final = "stretch"
STRETCH_USERNAME: Final = "stretch"
UNIT_LUMEN: Final = "lm"
USB: Final = "usb"

FLOW_NET: Final = "Network: Smile/Stretch"
FLOW_SMILE: Final = "Smile (Adam/Anna/P1)"
FLOW_STRETCH: Final = "Stretch (Stretch)"
FLOW_TYPE: Final = "flow_type"
FLOW_USB: Final = "USB: Stick"

UNDO_UPDATE_LISTENER: Final = "undo_update_listener"

# Default directives
DEFAULT_PORT: Final = 80
DEFAULT_SCAN_INTERVAL: Final[dict[str, timedelta]] = {
    "power": timedelta(seconds=10),
    "stretch": timedelta(seconds=60),
    "thermostat": timedelta(seconds=60),
}
DEFAULT_TIMEOUT: Final = 10
DEFAULT_USERNAME: Final = "smile"

# --- Const for Plugwise Smile and Stretch
PLATFORMS_GATEWAY: Final[list[str]] = [
    Platform.BINARY_SENSOR,
    Platform.CLIMATE,
    Platform.NUMBER,
    Platform.SELECT,
    Platform.SENSOR,
    Platform.SWITCH,
]
SENSOR_PLATFORMS: Final[list[str]] = [Platform.SENSOR, Platform.SWITCH]
SERVICE_DELETE: Final = "delete_notification"
SEVERITIES: Final[list[str]] = ["other", "info", "message", "warning", "error"]

# Climate const:
MASTER_THERMOSTATS: Final[list[str]] = [
    "thermostat",
    "zone_thermometer",
    "zone_thermostat",
    "thermostatic_radiator_valve",
]

# Config_flow const:
ZEROCONF_MAP: Final[dict[str, str]] = {
    "smile": "P1",
    "smile_thermo": "Anna",
    "smile_open_therm": "Adam",
    "stretch": "Stretch",
}

# Icons
COOLING_ICON: Final = "mdi:snowflake"
FLAME_ICON: Final = "mdi:fire"
FLOW_OFF_ICON: Final = "mdi:water-pump-off"
FLOW_ON_ICON: Final = "mdi:water-pump"
HEATING_ICON: Final = "mdi:radiator"
IDLE_ICON: Final = "mdi:circle-off-outline"
NOTIFICATION_ICON: Final = "mdi:mailbox-up-outline"
NO_NOTIFICATION_ICON: Final = "mdi:mailbox-outline"
SWITCH_ICON: Final = "mdi:electric-switch"

# Binary Sensors:
COMPRESSOR_STATE: Final = "compressor_state"
DHW_STATE: Final = "dhw_state"
FLAME_STATE: Final = "flame_state"
PW_NOTIFICATION: Final = "plugwise_notification"
SLAVE_BOILER_STATE: Final = "slave_boiler_state"

# Sensors:
BATTERY: Final = "battery"
CURRENT_TEMP: Final = "temperature"
DEVICE_STATE: Final = "device_state"
EL_CONSUMED: Final = "electricity_consumed"
EL_CONSUMED_INTERVAL: Final = "electricity_consumed_interval"
EL_CONSUMED_OFF_PEAK_CUMULATIVE: Final = "electricity_consumed_off_peak_cumulative"
EL_CONSUMED_OFF_PEAK_INTERVAL: Final = "electricity_consumed_off_peak_interval"
EL_CONSUMED_OFF_PEAK_POINT: Final = "electricity_consumed_off_peak_point"
EL_CONSUMED_PEAK_CUMULATIVE: Final = "electricity_consumed_peak_cumulative"
EL_CONSUMED_PEAK_INTERVAL: Final = "electricity_consumed_peak_interval"
EL_CONSUMED_PEAK_POINT: Final = "electricity_consumed_peak_point"
EL_CONSUMED_POINT: Final = "electricity_consumed_point"
EL_PRODUCED: Final = "electricity_produced"
EL_PRODUCED_INTERVAL: Final = "electricity_produced_interval"
EL_PRODUCED_OFF_PEAK_CUMULATIVE: Final = "electricity_produced_off_peak_cumulative"
EL_PRODUCED_OFF_PEAK_INTERVAL: Final = "electricity_produced_off_peak_interval"
EL_PRODUCED_OFF_PEAK_POINT: Final = "electricity_produced_off_peak_point"
EL_PRODUCED_PEAK_CUMULATIVE: Final = "electricity_produced_peak_cumulative"
EL_PRODUCED_PEAK_INTERVAL: Final = "electricity_produced_peak_interval"
EL_PRODUCED_PEAK_POINT: Final = "electricity_produced_peak_point"
EL_PRODUCED_POINT: Final = "electricity_produced_point"
GAS_CONSUMED_CUMULATIVE: Final = "gas_consumed_cumulative"
GAS_CONSUMED_INTERVAL: Final = "gas_consumed_interval"
HUMIDITY: Final = "humidity"
INTENDED_BOILER_TEMP: Final = "intended_boiler_temperature"
MOD_LEVEL: Final = "modulation_level"
NET_EL_CUMULATIVE: Final = "net_electricity_cumulative"
NET_EL_POINT: Final = "net_electricity_point"
OUTDOOR_AIR_TEMP: Final = "outdoor_air_temperature"
OUTDOOR_TEMP: Final = "outdoor_temperature"
RETURN_TEMP: Final = "return_temperature"
TARGET_TEMP: Final = "setpoint"
TARGET_TEMP_HIGH: Final = "setpoint_high"
TARGET_TEMP_LOW: Final = "setpoint_low"
TEMP_DIFF: Final = "temperature_difference"
VALVE_POS: Final = "valve_position"
WATER_PRESSURE: Final = "water_pressure"
WATER_TEMP: Final = "water_temperature"

# Switches
DHW_COMF_MODE: Final = "dhw_cm_switch"
LOCK: Final = "lock"
RELAY: Final = "relay"


# --- Const for Plugwise USB-stick.

PLATFORMS_USB: Final[list[str]] = [
    Platform.BINARY_SENSOR,
    Platform.SENSOR,
    Platform.SWITCH,
]
CONF_USB_PATH: Final = "usb_path"

# Callback types
CB_NEW_NODE: Final = "NEW_NODE"
CB_JOIN_REQUEST: Final = "JOIN_REQUEST"


# USB generic device constants
USB_AVAILABLE_ID: Final = "available"

ATTR_MAC_ADDRESS: Final = "mac"

SERVICE_USB_DEVICE_ADD: Final = "device_add"
SERVICE_USB_DEVICE_REMOVE: Final = "device_remove"
SERVICE_USB_DEVICE_SCHEMA: Final = vol.Schema(
    {vol.Required(ATTR_MAC_ADDRESS): cv.string}
)


# USB Relay device constants
USB_RELAY_ID: Final = "relay"


# USB SED (battery powered) device constants
ATTR_SED_STAY_ACTIVE: Final = "stay_active"
ATTR_SED_SLEEP_FOR: Final = "sleep_for"
ATTR_SED_MAINTENANCE_INTERVAL: Final = "maintenance_interval"
ATTR_SED_CLOCK_SYNC: Final = "clock_sync"
ATTR_SED_CLOCK_INTERVAL: Final = "clock_interval"

SERVICE_USB_SED_BATTERY_CONFIG: Final = "configure_battery_savings"
SERVICE_USB_SED_BATTERY_CONFIG_SCHEMA: Final = {
    vol.Required(ATTR_SED_STAY_ACTIVE): vol.All(
        vol.Coerce(int), vol.Range(min=1, max=120)
    ),
    vol.Required(ATTR_SED_SLEEP_FOR): vol.All(
        vol.Coerce(int), vol.Range(min=10, max=60)
    ),
    vol.Required(ATTR_SED_MAINTENANCE_INTERVAL): vol.All(
        vol.Coerce(int), vol.Range(min=5, max=1440)
    ),
    vol.Required(ATTR_SED_CLOCK_SYNC): cv.boolean,
    vol.Required(ATTR_SED_CLOCK_INTERVAL): vol.All(
        vol.Coerce(int), vol.Range(min=60, max=10080)
    ),
}


# USB Scan device constants
USB_MOTION_ID: Final = "motion"

ATTR_SCAN_DAYLIGHT_MODE: Final = "day_light"
ATTR_SCAN_SENSITIVITY_MODE: Final = "sensitivity_mode"
ATTR_SCAN_RESET_TIMER: Final = "reset_timer"

SCAN_SENSITIVITY_HIGH: Final = "high"
SCAN_SENSITIVITY_MEDIUM: Final = "medium"
SCAN_SENSITIVITY_OFF: Final = "off"
SCAN_SENSITIVITY_MODES = [
    SCAN_SENSITIVITY_HIGH,
    SCAN_SENSITIVITY_MEDIUM,
    SCAN_SENSITIVITY_OFF,
]

SERVICE_USB_SCAN_CONFIG: Final = "configure_scan"
SERVICE_USB_SCAN_CONFIG_SCHEMA = (
    {
        vol.Required(ATTR_SCAN_SENSITIVITY_MODE): vol.In(SCAN_SENSITIVITY_MODES),
        vol.Required(ATTR_SCAN_RESET_TIMER): vol.All(
            vol.Coerce(int), vol.Range(min=1, max=240)
        ),
        vol.Required(ATTR_SCAN_DAYLIGHT_MODE): cv.boolean,
    },
)
