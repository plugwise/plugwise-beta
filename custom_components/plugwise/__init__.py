"""Plugwise platform for Home Assistant Core."""

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_HOST
from homeassistant.core import HomeAssistant

from .gateway import async_setup_entry_gw, async_unload_entry_gw


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up Plugwise components from a config entry."""
    if entry.data.get(CONF_HOST):
        return await async_setup_entry_gw(hass, entry)
    # PLACEHOLDER USB entry setup
    return False


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload the Plugwise components."""
    if entry.data.get(CONF_HOST):
        return await async_unload_entry_gw(hass, entry)
    # PLACEHOLDER USB entry setup
    return False
