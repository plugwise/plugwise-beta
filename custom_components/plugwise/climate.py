"""Plugwise Climate component for Home Assistant."""
from __future__ import annotations

from typing import Any

from homeassistant.components.climate import ClimateEntity
from homeassistant.components.climate.const import (
    CURRENT_HVAC_COOL,
    CURRENT_HVAC_HEAT,
    CURRENT_HVAC_IDLE,
    HVAC_MODE_AUTO,
    HVAC_MODE_HEAT,
    HVAC_MODE_COOL,
    HVAC_MODE_OFF,
    PRESET_AWAY,  # pw-beta homekit emulation
    PRESET_HOME,  # pw-beta homekit emulation
    SUPPORT_PRESET_MODE,
    SUPPORT_TARGET_TEMPERATURE,
)
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import ATTR_TEMPERATURE, TEMP_CELSIUS
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from .const import (
    CONF_HOMEKIT_EMULATION,  # pw-beta homekit emulation
    COORDINATOR,
    DEFAULT_MAX_TEMP,
    DEFAULT_MIN_TEMP,
    DOMAIN,
    MASTER_THERMOSTATS,
)
from .coordinator import PlugwiseDataUpdateCoordinator
from .entity import PlugwiseEntity
from .util import plugwise_command


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up the Smile Thermostats from a config entry."""
    coordinator = hass.data[DOMAIN][config_entry.entry_id][COORDINATOR]

    # pw-beta homekit emulation
    homekit_enabled = config_entry.options.get(CONF_HOMEKIT_EMULATION, False)

    async_add_entities(
        PlugwiseClimateEntity(coordinator, device_id, homekit_enabled)
        for device_id, device in coordinator.data.devices.items()
        if device["class"] in MASTER_THERMOSTATS
    )


class PlugwiseClimateEntity(PlugwiseEntity, ClimateEntity):
    """Representation of an Plugwise thermostat."""

    _attr_temperature_unit = TEMP_CELSIUS

    def __init__(
        self,
        coordinator: PlugwiseDataUpdateCoordinator,
        device_id: str,
        enabled: bool,  # pw-beta homekit emulation
    ) -> None:
        """Set up the Plugwise API."""
        super().__init__(coordinator, device_id)
        self._homekit_enabled = enabled  # pw-beta homekit emulation
        self._homekit_mode: str | None = None  # pw-beta homekit emulation
        self._attr_extra_state_attributes = {}
        self._attr_unique_id = f"{device_id}-climate"
        self._attr_name = self.device.get("name")

        # Determine preset modes
        self._attr_supported_features = SUPPORT_TARGET_TEMPERATURE
        if presets := self.device.get("presets"):
            self._attr_supported_features |= SUPPORT_PRESET_MODE
            self._attr_preset_modes = list(presets)

        # Determine hvac modes and current hvac mode
        self._attr_hvac_modes = [HVAC_MODE_HEAT]
        if self.coordinator.data.gateway.get("cooling_present"):
            self._attr_hvac_modes.append(HVAC_MODE_COOL)
        if self.device.get("available_schedules") != ["None"]:
            self._attr_hvac_modes.append(HVAC_MODE_AUTO)
        if self._homekit_enabled:  # pw-beta homekit emulation
            self._attr_hvac_modes.append(HVAC_MODE_OFF)

        self._attr_min_temp = self.device.get("lower_bound", DEFAULT_MIN_TEMP)
        self._attr_max_temp = self.device.get("upper_bound", DEFAULT_MAX_TEMP)
        if resolution := self.device.get("resolution"):
            # Ensure we don't drop below 0.1
            self._attr_target_temperature_step = max(resolution, 0.1)

    @property
    def current_temperature(self) -> float | None:
        """Return the current temperature."""
        return self.device["sensors"].get("temperature")

    @property
    def target_temperature(self) -> float | None:
        """Return the temperature we try to reach."""
        return self.device["sensors"].get("setpoint")

    @property
    def hvac_mode(self) -> str:
        """Return HVAC operation ie. auto, heat, cool, or off mode."""
        if (mode := self.device.get("mode")) is None or mode not in self.hvac_modes:
            return HVAC_MODE_HEAT
        # pw-beta homekit emulation
        if (
            self._homekit_enabled and self._homekit_mode == HVAC_MODE_OFF
        ):
            mode = HVAC_MODE_OFF
        return mode

    @property
    def hvac_action(self) -> str:
        """Return the current running hvac operation if supported."""
        # When control_state is present, prefer this data
        if "control_state" in self.device:
            if self.device.get("control_state") == "cooling":
                return CURRENT_HVAC_COOL
            # Support preheating state as heating, until preheating is added as a separate state
            if self.device.get("control_state") in ["heating", "preheating"]:
                return CURRENT_HVAC_HEAT
        else:
            heater_central_data = self.coordinator.data.devices[
                self.coordinator.data.gateway["heater_id"]
            ]
            if heater_central_data["binary_sensors"].get("heating_state"):
                return CURRENT_HVAC_HEAT
            if heater_central_data["binary_sensors"].get("cooling_state"):
                return CURRENT_HVAC_COOL
        return CURRENT_HVAC_IDLE

    @property
    def preset_mode(self) -> str | None:
        """Return the current preset mode."""
        return self.device.get("active_preset")

    @plugwise_command
    async def async_set_temperature(self, **kwargs: Any) -> None:
        """Set new target temperature."""
        if ((temperature := kwargs.get(ATTR_TEMPERATURE)) is None) or (
            self._attr_max_temp < temperature < self._attr_min_temp
        ):
            raise ValueError("Invalid temperature requested")
        await self.coordinator.api.set_temperature(self.device["location"], temperature)

    @plugwise_command
    async def async_set_hvac_mode(self, hvac_mode: str) -> None:
        """Set the hvac mode."""
        self._homekit_mode = hvac_mode  # pw-beta homekit emulation
        if hvac_mode == HVAC_MODE_AUTO and not self.device.get("schedule_temperature"):
            raise ValueError("Cannot set HVAC mode to Auto: No schedule available")

        await self.coordinator.api.set_schedule_state(
            self.device["location"],
            self.device.get("last_used"),
            "on" if hvac_mode == HVAC_MODE_AUTO else "off",
        )

        # pw-beta: feature request - mimic HomeKit behavior
        if self._homekit_enabled:
            if self._homekit_mode == HVAC_MODE_OFF:
                await self.async_set_preset_mode(PRESET_AWAY)
            if (
                self._homekit_mode in [HVAC_MODE_HEAT, HVAC_MODE_COOL]
                and self.device["active_preset"] == PRESET_AWAY
            ):
                await self.async_set_preset_mode(PRESET_HOME)

    @plugwise_command
    async def async_set_preset_mode(self, preset_mode: str) -> None:
        """Set the preset mode."""
        await self.coordinator.api.set_preset(self.device["location"], preset_mode)
