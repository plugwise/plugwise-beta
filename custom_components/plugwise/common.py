"""Common functions for Plugwise integration."""

from homeassistant.core import HomeAssistant

from .const import COORDINATOR, DOMAIN
from .coordinator import PlugwiseDataUpdateCoordinator


def get_coordinator(
    hass: HomeAssistant, config_entry_id: str
) -> PlugwiseDataUpdateCoordinator:
    """Get coordinator for given config entry id."""
    coordinator: PlugwiseDataUpdateCoordinator = hass.data[DOMAIN][config_entry_id][
        COORDINATOR
    ]

    return coordinator
