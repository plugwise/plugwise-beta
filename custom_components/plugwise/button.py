"""Plugwise Button component for Home Assistant."""
from __future__ import annotations

from plugwise import Smile

from homeassistant.components.button import (
    ButtonDeviceClass,
    ButtonEntity,
    ButtonEntityDescription,
)
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import ATTR_NAME, EntityCategory, UnitOfTemperature
from homeassistant.core import HomeAssistant, callback
from homeassistant.helpers.entity_platform import AddEntitiesCallback

from .const import GATEWAY_ID, LOGGER
from .coordinator import PlugwiseDataUpdateCoordinator
from .entity import get_coordinator


BUTTON_TYPES: tuple[ButtonEntityDescription, ...] = (
    ButtonEntityDescription(
        key="reboot",
        translation_key="reboot"
        device_class=ButtonDeviceClass.RESTART,
        entity_category=EntityCategory.CONFIG,
    )
)


 async def async_setup_entry(
    hass: HomeAssistant,
    entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up the Plugwise buttons from a ConfigEntry."""
    coordinator = get_coordinator(hass, entry.entry_id)

    @callback
    def _add_entities() -> None:
        """Add Entities."""
        if not coordinator.new_devices:
            return

        entities: list[PlugwiseButtonEntity] = []
        for device_id, device in coordinator.data.devices.items():
            if device_id == coordinator.data.gateway[GATEWAY_ID]:
                for description in BUTTON_TYPES:
                    entities.append(
                        PlugwiseButtonEntity(
                            coordinator,
                            device_id,
                            description,
                        )
                )
                LOGGER.debug(
                    "Add %s %s button", device[ATTR_NAME], description.key
                )

        async_add_entities(entities)

    entry.async_on_unload(coordinator.async_add_listener(_add_entities))

    _add_entities()


class PlugwiseButton(ButtonEntity):
    """Defines a Plugwise button."""

    entity_description: ButtonEntityDescription

    def __init__(
        self,
        coordinator: PlugwiseDataUpdateCoordinator,
        device_id: str,
        description: PlugwiseButtonEntityDescription,
    ) -> None:
        """Initialize the button."""
        super().__init__(coordinator, device_id)
        self.device_id = device_id
        self.entity_description = description
        self._attr_unique_id = f"{device_id}-{description.key}"

    @plugwise_command
    async def async_press(self) -> None:
        """Triggers the Shelly button press service."""
        await self.coordinator.api.reboot_gateway()
