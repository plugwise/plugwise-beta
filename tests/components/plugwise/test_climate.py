"""Tests for the Plugwise Climate integration."""

from unittest.mock import MagicMock

from plugwise.exceptions import PlugwiseException
import pytest

from homeassistant.components.climate.const import (
    HVAC_MODE_AUTO,
    HVAC_MODE_COOL,
    HVAC_MODE_HEAT,
    HVAC_MODE_HEAT_COOL,
)
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import HomeAssistantError

from tests.common import MockConfigEntry

TEST_HOST = "1.1.1.1"
TEST_PASSWORD = "test_password"


async def test_adam_climate_entity_attributes(
    hass: HomeAssistant, mock_smile_adam: MagicMock, init_integration: MockConfigEntry
) -> None:
    """Test creation of adam climate device environment."""
    state = hass.states.get("climate.zone_lisa_wk")
    assert state
    assert state.state == HVAC_MODE_AUTO
    assert state.attributes["hvac_modes"] == [HVAC_MODE_HEAT, HVAC_MODE_AUTO]
    # hvac_action is not asserted as the fixture is not in line with recent firmware functionality

    assert "preset_modes" in state.attributes
    assert "no_frost" in state.attributes["preset_modes"]
    assert "home" in state.attributes["preset_modes"]

    assert state.attributes["current_temperature"] == 20.9
    assert state.attributes["preset_mode"] == "home"
    assert state.attributes["supported_features"] == 17
    assert state.attributes["temperature"] == 21.5
    assert state.attributes["min_temp"] == 0.0
    assert state.attributes["max_temp"] == 99.9
    assert state.attributes["target_temp_step"] == 0.1

    state = hass.states.get("climate.zone_thermostat_jessie")
    assert state
    assert state.state == HVAC_MODE_AUTO
    assert state.attributes["hvac_modes"] == [HVAC_MODE_HEAT, HVAC_MODE_AUTO]
    # hvac_action is not asserted as the fixture is not in line with recent firmware functionality

    assert "preset_modes" in state.attributes
    assert "no_frost" in state.attributes["preset_modes"]
    assert "home" in state.attributes["preset_modes"]

    assert state.attributes["current_temperature"] == 17.2
    assert state.attributes["preset_mode"] == "asleep"
    assert state.attributes["temperature"] == 15.0
    assert state.attributes["min_temp"] == 0.0
    assert state.attributes["max_temp"] == 99.9
    assert state.attributes["target_temp_step"] == 0.1


async def test_adam_2_climate_entity_attributes(
    hass: HomeAssistant, mock_smile_adam_2: MagicMock, init_integration: MockConfigEntry
) -> None:
    """Test creation of adam climate device environment."""
    state = hass.states.get("climate.anna")
    assert state
    assert state.state == HVAC_MODE_HEAT
    assert state.attributes["hvac_action"] == "heating"
    assert state.attributes["hvac_modes"] == [HVAC_MODE_HEAT, HVAC_MODE_AUTO]

    state = hass.states.get("climate.lisa_badkamer")
    assert state
    assert state.state == HVAC_MODE_AUTO
    assert state.attributes["hvac_action"] == "idle"
    assert state.attributes["hvac_modes"] == [HVAC_MODE_HEAT, HVAC_MODE_AUTO]


async def test_adam_3_climate_entity_attributes(
    hass: HomeAssistant, mock_smile_adam_3: MagicMock, init_integration: MockConfigEntry
) -> None:
    """Test creation of adam climate device environment."""
    state = hass.states.get("climate.anna")

    assert state
    assert state.state == HVAC_MODE_COOL
    assert state.attributes["hvac_action"] == "cooling"
    assert state.attributes["hvac_modes"] == [HVAC_MODE_COOL, HVAC_MODE_AUTO]


async def test_adam_climate_adjust_negative_testing(
    hass: HomeAssistant, mock_smile_adam: MagicMock, init_integration: MockConfigEntry
) -> None:
    """Test exceptions of climate entities."""
    mock_smile_adam.set_preset.side_effect = PlugwiseException
    mock_smile_adam.set_schedule_state.side_effect = PlugwiseException
    mock_smile_adam.set_temperature.side_effect = PlugwiseException

    with pytest.raises(HomeAssistantError):
        await hass.services.async_call(
            "climate",
            "set_temperature",
            {"entity_id": "climate.zone_lisa_wk", "temperature": 25},
            blocking=True,
        )

    with pytest.raises(HomeAssistantError):
        await hass.services.async_call(
            "climate",
            "set_preset_mode",
            {"entity_id": "climate.zone_thermostat_jessie", "preset_mode": "home"},
            blocking=True,
        )


async def test_adam_climate_entity_climate_changes(
    hass: HomeAssistant, mock_smile_adam: MagicMock, init_integration: MockConfigEntry
) -> None:
    """Test handling of user requests in adam climate device environment."""
    await hass.services.async_call(
        "climate",
        "set_temperature",
        {"entity_id": "climate.zone_lisa_wk", "temperature": 25},
        blocking=True,
    )

    assert mock_smile_adam.set_temperature.call_count == 1
    mock_smile_adam.set_temperature.assert_called_with(
        "c50f167537524366a5af7aa3942feb1e", {"setpoint": 25.0}
    )

    with pytest.raises(ValueError):
        await hass.services.async_call(
            "climate",
            "set_temperature",
            {"entity_id": "climate.zone_lisa_wk", "temperature": 150},
            blocking=True,
        )

    with pytest.raises(HomeAssistantError):
        await hass.services.async_call(
            "climate",
            "set_temperature",
            {
                "entity_id": "climate.zone_lisa_wk",
                "hvac_mode": "heat",
                "temperature": 25,
            },
            blocking=True,
        )

    await hass.services.async_call(
        "climate",
        "set_preset_mode",
        {"entity_id": "climate.zone_lisa_wk", "preset_mode": "away"},
        blocking=True,
    )

    assert mock_smile_adam.set_preset.call_count == 1
    mock_smile_adam.set_preset.assert_called_with(
        "c50f167537524366a5af7aa3942feb1e", "away"
    )

    await hass.services.async_call(
        "climate",
        "set_temperature",
        {"entity_id": "climate.zone_thermostat_jessie", "temperature": 25},
        blocking=True,
    )

    assert mock_smile_adam.set_temperature.call_count == 2
    mock_smile_adam.set_temperature.assert_called_with(
        "82fa13f017d240daa0d0ea1775420f24", {"setpoint": 25.0}
    )

    await hass.services.async_call(
        "climate",
        "set_preset_mode",
        {"entity_id": "climate.zone_thermostat_jessie", "preset_mode": "home"},
        blocking=True,
    )

    assert mock_smile_adam.set_preset.call_count == 2
    mock_smile_adam.set_preset.assert_called_with(
        "82fa13f017d240daa0d0ea1775420f24", "home"
    )

    with pytest.raises(HomeAssistantError):
        await hass.services.async_call(
            "climate",
            "set_hvac_mode",
            {
                "entity_id": "climate.zone_thermostat_jessie",
                "hvac_mode": "dry",
            },
            blocking=True,
        )


async def test_anna_climate_entity_attributes(
    hass: HomeAssistant, mock_smile_anna: MagicMock, init_integration: MagicMock
) -> None:
    """Test creation of anna climate device environment."""
    state = hass.states.get("climate.anna")
    assert state
    assert state.state == HVAC_MODE_AUTO
    assert state.attributes["hvac_action"] == "heating"
    assert state.attributes["hvac_modes"] == [
        HVAC_MODE_HEAT_COOL,
        HVAC_MODE_AUTO,
    ]

    assert "no_frost" in state.attributes["preset_modes"]
    assert "home" in state.attributes["preset_modes"]

    assert state.attributes["current_temperature"] == 19.3
    assert state.attributes["preset_mode"] == "home"
    assert state.attributes["supported_features"] == 18
    assert state.attributes["target_temp_high"] == 24.0
    assert state.attributes["target_temp_low"] == 21.0
    assert state.attributes["min_temp"] == 4.0
    assert state.attributes["max_temp"] == 30.0
    assert state.attributes["target_temp_step"] == 0.1


async def test_anna_2_climate_entity_attributes(
    hass: HomeAssistant,
    mock_smile_anna_2: MagicMock,
    init_integration: MockConfigEntry,
) -> None:
    """Test creation of anna climate device environment."""
    state = hass.states.get("climate.anna")
    assert state
    assert state.state == HVAC_MODE_AUTO
    assert state.attributes["hvac_action"] == "cooling"
    assert state.attributes["hvac_modes"] == [
        HVAC_MODE_HEAT_COOL,
        HVAC_MODE_AUTO,
    ]
    assert state.attributes["target_temp_high"] == 24.0
    assert state.attributes["target_temp_low"] == 20.5
    assert state.attributes["supported_features"] == 18


async def test_anna_3_climate_entity_attributes(
    hass: HomeAssistant,
    mock_smile_anna_3: MagicMock,
    init_integration: MockConfigEntry,
) -> None:
    """Test creation of anna climate device environment."""
    state = hass.states.get("climate.anna")
    assert state
    assert state.state == HVAC_MODE_AUTO
    assert state.attributes["hvac_action"] == "idle"
    assert state.attributes["hvac_modes"] == [
        HVAC_MODE_HEAT_COOL,
        HVAC_MODE_AUTO,
    ]


async def test_anna_climate_entity_climate_changes(
    hass: HomeAssistant, mock_smile_anna: MagicMock, init_integration: MockConfigEntry
) -> None:
    """Test handling of user requests in anna climate device environment."""
    await hass.services.async_call(
        "climate",
        "set_temperature",
        {"entity_id": "climate.anna", "target_temp_high": 25, "target_temp_low": 20},
        blocking=True,
    )

    assert mock_smile_anna.set_temperature.call_count == 1
    mock_smile_anna.set_temperature.assert_called_with(
        "c784ee9fdab44e1395b8dee7d7a497d5",
        {"setpoint_high": 25.0, "setpoint_low": 20.0},
    )

    await hass.services.async_call(
        "climate",
        "set_preset_mode",
        {"entity_id": "climate.anna", "preset_mode": "away"},
        blocking=True,
    )

    assert mock_smile_anna.set_preset.call_count == 1
    mock_smile_anna.set_preset.assert_called_with(
        "c784ee9fdab44e1395b8dee7d7a497d5", "away"
    )

    await hass.services.async_call(
        "climate",
        "set_hvac_mode",
        {"entity_id": "climate.anna", "hvac_mode": "heat_cool"},
        blocking=True,
    )

    assert mock_smile_anna.set_temperature.call_count == 1
    assert mock_smile_anna.set_schedule_state.call_count == 1
    mock_smile_anna.set_schedule_state.assert_called_with(
        "c784ee9fdab44e1395b8dee7d7a497d5", "standaard", "off"
    )

    await hass.services.async_call(
        "climate",
        "set_hvac_mode",
        {"entity_id": "climate.anna", "hvac_mode": "auto"},
        blocking=True,
    )

    assert mock_smile_anna.set_schedule_state.call_count == 2
    mock_smile_anna.set_schedule_state.assert_called_with(
        "c784ee9fdab44e1395b8dee7d7a497d5", "standaard", "on"
    )
