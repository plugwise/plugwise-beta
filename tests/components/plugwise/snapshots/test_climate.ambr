# serializer version: 1
# name: test_binary_sensor_states[platforms0-True-anna_heatpump_heating][climate.anna-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': dict({
      'hvac_modes': list([
        <HVACMode.AUTO: 'auto'>,
        <HVACMode.HEAT_COOL: 'heat_cool'>,
      ]),
      'max_temp': 30.0,
      'min_temp': 4.0,
      'preset_modes': list([
        'no_frost',
        'home',
        'away',
        'asleep',
        'vacation',
      ]),
      'target_temp_step': 0.1,
    }),
    'config_entry_id': <ANY>,
    'config_subentry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'climate',
    'entity_category': None,
    'entity_id': 'climate.anna',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': None,
    'platform': 'plugwise',
    'previous_unique_id': None,
    'suggested_object_id': None,
    'supported_features': <ClimateEntityFeature: 18>,
    'translation_key': 'plugwise',
    'unique_id': '3cb70739631c4d17a86b8b12e8a5161b-climate',
    'unit_of_measurement': None,
  })
# ---
# name: test_binary_sensor_states[platforms0-True-anna_heatpump_heating][climate.anna-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'current_temperature': 19.3,
      'friendly_name': 'Anna',
      'hvac_action': <HVACAction.HEATING: 'heating'>,
      'hvac_modes': list([
        <HVACMode.AUTO: 'auto'>,
        <HVACMode.HEAT_COOL: 'heat_cool'>,
      ]),
      'max_temp': 30.0,
      'min_temp': 4.0,
      'preset_mode': 'home',
      'preset_modes': list([
        'no_frost',
        'home',
        'away',
        'asleep',
        'vacation',
      ]),
      'supported_features': <ClimateEntityFeature: 18>,
      'target_temp_high': 30.0,
      'target_temp_low': 20.5,
      'target_temp_step': 0.1,
    }),
    'context': <ANY>,
    'entity_id': 'climate.anna',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'auto',
  })
# ---
